File: backend/src/models/action_queue.py
"""Action queue Pydantic models for ARIA (US-937).

This module contains models for the autonomous action queue and approval workflow.
"""

from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


class ActionAgent(str, Enum):
    """Agent that initiated the action."""

    SCOUT = "scout"
    ANALYST = "analyst"
    HUNTER = "hunter"
    OPERATOR = "operator"
    SCRIBE = "scribe"
    STRATEGIST = "strategist"


class ActionType(str, Enum):
    """Type of autonomous action."""

    EMAIL_DRAFT = "email_draft"
    CRM_UPDATE = "crm_update"
    RESEARCH = "research"
    MEETING_PREP = "meeting_prep"
    LEAD_GEN = "lead_gen"


class RiskLevel(str, Enum):
    """Risk level determining approval workflow."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class ExecutionMode(str, Enum):
    """Execution mode determined by trust × risk matrix."""

    AUTO_EXECUTE = "auto_execute"
    EXECUTE_AND_NOTIFY = "execute_and_notify"
    APPROVE_PLAN = "approve_plan"
    APPROVE_EACH = "approve_each"


class ActionStatus(str, Enum):
    """Current status of an action."""

    PENDING = "pending"
    APPROVED = "approved"
    AUTO_APPROVED = "auto_approved"
    EXECUTING = "executing"
    COMPLETED = "completed"
    REJECTED = "rejected"
    FAILED = "failed"
    UNDO_PENDING = "undo_pending"


class ActionCreate(BaseModel):
    """Request model for submitting a new action."""

    agent: ActionAgent
    action_type: ActionType
    title: str
    description: str | None = None
    risk_level: RiskLevel
    payload: dict[str, Any] = Field(default_factory=dict)
    reasoning: str | None = None
    risk_score: float | None = None
    task_characteristics: dict[str, Any] | None = None


class ActionReject(BaseModel):
    """Request model for rejecting an action."""

    reason: str | None = None


class BatchApproveRequest(BaseModel):
    """Request model for batch-approving actions."""

    action_ids: list[str]

File: frontend/src/api/actionQueue.ts
import { apiClient } from "./client";

// Enums matching backend
export type ActionAgent = "scout" | "analyst" | "hunter" | "operator" | "scribe" | "strategist";
export type ActionType = "email_draft" | "crm_update" | "research" | "meeting_prep" | "lead_gen";
export type RiskLevel = "low" | "medium" | "high" | "critical";
export type ActionStatus =
  | "pending"
  | "approved"
  | "auto_approved"
  | "executing"
  | "completed"
  | "rejected"
  | "failed"
  | "undo_pending";

// Response types
export interface Action {
  id: string;
  user_id: string;
  agent: ActionAgent;
  action_type: ActionType;
  title: string;
  description: string | null;
  risk_level: RiskLevel;
  status: ActionStatus;
  payload: Record<string, unknown>;
  reasoning: string | null;
  result: Record<string, unknown>;
  approved_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

// Request types
export interface SubmitActionData {
  agent: ActionAgent;
  action_type: ActionType;
  title: string;
  description?: string;
  risk_level: RiskLevel;
  payload?: Record<string, unknown>;
  reasoning?: string;
}

export interface BatchApproveResponse {
  approved: Action[];
  count: number;
}

// API functions
export async function listActions(status?: ActionStatus, limit?: number): Promise<Action[]> {
  const params = new URLSearchParams();
  if (status) params.append("status", status);
  if (limit) params.append("limit", limit.toString());

  const url = params.toString() ? `/actions?${params}` : "/actions";
  const response = await apiClient.get<Action[]>(url);
  return response.data;
}

export async function getAction(actionId: string): Promise<Action> {
  const response = await apiClient.get<Action>(`/actions/${actionId}`);
  return response.data;
}

export async function submitAction(data: SubmitActionData): Promise<Action> {
  const response = await apiClient.post<Action>("/actions", data);
  return response.data;
}

export async function approveAction(actionId: string): Promise<Action> {
  const response = await apiClient.post<Action>(`/actions/${actionId}/approve`);
  return response.data;
}

export async function rejectAction(actionId: string, reason?: string): Promise<Action> {
  const response = await apiClient.post<Action>(`/actions/${actionId}/reject`, {
    reason: reason ?? null,
  });
  return response.data;
}

export async function batchApproveActions(actionIds: string[]): Promise<BatchApproveResponse> {
  const response = await apiClient.post<BatchApproveResponse>("/actions/batch-approve", {
    action_ids: actionIds,
  });
  return response.data;
}

export async function getPendingCount(): Promise<number> {
  const response = await apiClient.get<{ count: number }>("/actions/pending-count");
  return response.data.count;
}

export async function executeAction(actionId: string): Promise<Action> {
  const response = await apiClient.post<Action>(`/actions/${actionId}/execute`);
  return response.data;
}

export interface UndoResponse {
  success: boolean;
  action_id: string;
  reversal?: Record<string, unknown>;
  reason?: string;
}

export async function undoAction(actionId: string): Promise<UndoResponse> {
  const response = await apiClient.post<UndoResponse>(`/actions/${actionId}/undo`);
  return response.data;
}
